### DISCUSSION TITLE
Target creation using constructor

### DISCUSSION THREAD

**Discussion by abhishekgoyal1994** [mapstruct/mapstruct#2761] [Q&A]
How does mapstruct decides between the following approaches:

* Get all parameters using getters of source and then call constructor of target to create the mapping.
* Call constructor of target with all null parameters, then set each of them using setters.

```
Target target(Source source) {
  if ( source == null ) {
      return null;
  }

  String a = null;
  String b = null;
  String c = null;

  a = source.getA();
  b = source.getB();
  c = source.getC();

  Target target = new Target( a, b, c );

  return target;
}
```

```
Target target(Source source) {
    if ( source == null ) {
        return null;
    }

    String a = null;
    String b = null;
    String c = null;

    Target target = new Target( a, b, c );

    target.setA(source.getA());
    target.setA(source.getB());
    target.setA(source.getC());

    return target;
}
```

**Comment by Zegveld** [comment]
If you always pick the pass in null values, then you run into an unwanted situation when the constructor does not allow null-values.  Seeing as that is not wanted, mapstruct will always attempt to initialize the properties used in the constructor instead of putting in null values first and then calling the setters.
